src/timed.rs

```diff
@@ -0,0 +1,220 @@
+    }
+}
+
+impl<M: Send + 'static, A: Actor<Context = TimedContext<M>, Message = M>> Actor for Timed<A> {
+    type Context = Context<Self::Message>;
+    type Error = A::Error;
+    type Message = TimedMessage<M>;
+
+    fn handle(
+        &mut self,
+        context: &mut Self::Context,
+        timed_message: Self::Message,
+    ) -> Result<(), Self::Error> {
+        let item = match timed_message {
+            TimedMessage::Instant { message } => {
+                return self.inner.handle(&mut TimedContext::from_context(context), message);
+            },
+            TimedMessage::Delayed { message, fire_at } => {
+                QueueItem { fire_at, payload: Payload::Delayed { message } }
+            },
+            TimedMessage::Recurring { factory, fire_at, interval } => {
+                QueueItem { fire_at, payload: Payload::Recurring { factory, interval } }
+            },
+        };
+
+        self.queue.push(item);
+        self.schedule_timeout(context);
+        Ok(())
+    }
+
+    fn name() -> &'static str {
+        A::name()
+    }
+
+    fn started(&mut self, context: &mut Self::Context) {
+        self.inner.started(&mut TimedContext::from_context(context))
+    }
+
+    fn stopped(&mut self, context: &mut Self::Context) {
+        self.inner.stopped(&mut TimedContext::from_context(context))
+    }
+
+    fn deadline_passed(
+        &mut self,
+        context: &mut Self::Context,
+        _deadline: Instant,
+    ) -> Result<(), Self::Error> {
+        // Handle all messages that should have been handled by now.
+        let now = Instant::now();
+        while self.queue.peek().map(|m| m.fire_at <= now).unwrap_or(false) {
+            let item = self.queue.pop().expect("heap is non-empty, we have just peeked");
+
+            let message = match item.payload {
+                Payload::Delayed { message } => message,
+                Payload::Recurring { mut factory, interval } => {
+                    let message = factory();
+                    self.queue.push(QueueItem {
+                        fire_at: item.fire_at + interval,
+                        payload: Payload::Recurring { factory, interval },
+                    });
+                    message
+                },
+            };
+
+            // Let inner actor do its job.
+            self.inner.handle(&mut TimedContext::from_context(context), message)?;
+        }
+
+        self.schedule_timeout(context);
+        Ok(())
+    }
+}
+
+/// Access wrapped actor.
+impl<A: Actor> Deref for Timed<A> {
+    type Target = A;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+/// Implementation detail, element of message queue ordered by time to fire at.
+struct QueueItem<M> {
+    fire_at: Instant,
+    payload: Payload<M>,
+}
+
+impl<M> PartialEq for QueueItem<M> {
+    fn eq(&self, other: &Self) -> bool {
+        self.fire_at == other.fire_at
```

Comment:
Is it okay that the `payload` field isn't considered here? If you scheduled two messages to fire at the same by reusing some `Instant`, would they both fire?

Reply:
That should be safe: we never compare queue items for equality ourselves and we never deduplicate them any any way. `PartialEq` and friends are just a requirement of `BinaryHeap`.

I wish there existed some Python-style `BinaryHeap::with_key_func<K: Ord>(f: Fn(V) -> K)` so that we can define the comparison only in the context of the container and don't mess with global `{Partial,}{Eq,Ord}`.